File: IFATCCompanion/AirportSurface/AirportSurfaceDiagnostics.swift
import Foundation

/// A point-in-time snapshot of the airport-surface feature for the Airport Surface
/// Diagnostics view and its text export. Always identifies OpenStreetMap / ODbL 1.0 and
/// carries the visible attribution.
struct AirportSurfaceDiagnostics {
    var airportID: String
    var sourceProvider: String
    var license: String
    var attribution: String
    var endpoint: String
    var fetchDate: Date?
    var cacheAgeDays: Int?
    var stale: Bool
    var rawFeatureCount: Int
    var runwayCount: Int
    var taxiwayCount: Int
    var taxilaneCount: Int
    var holdingPositionCount: Int
    var parkingCount: Int
    var apronCount: Int
    var graphNodeCount: Int
    var graphEdgeCount: Int
    var disconnectedComponents: Int
    var inferredConnectors: Int
    var snappedSegment: String
    var routeSummary: String
    var routeDistanceMeters: Double?
    var runwayCrossings: Int
    var routeConfidence: SurfaceConfidence
    var datasetConfidence: SurfaceConfidence
    var nextCrossing: String
    var crossingState: String
    var authorizationState: String
    var statusText: String
    var lastError: String?

    init(surface: AirportSurfaceModel?, graph: SurfaceGraph?, route: SurfaceTaxiRoute?,
         kind: TaxiKind, status: AirportSurfaceStatus, datasetConfidence: SurfaceConfidence,
         routeConfidence: SurfaceConfidence, crossingState: RunwayCrossingState,
         activeCrossing: RouteCrossing?, progress: RouteTracker.Progress?,
         awaitingCrossingReadback: Bool, authorizedCrossingIndex: Int?,
         snappedSegment: String, lastError: String?) {

        airportID = surface?.icao ?? "—"
        sourceProvider = OSMSurface.providerName
        license = OSMSurface.licenseShortName
        attribution = OSMSurface.attributionText
        endpoint = surface?.source.endpoint ?? OSMSurface.primaryOverpassEndpoint
        fetchDate = surface?.source.fetchDate
        cacheAgeDays = surface.map { $0.source.cacheAgeDays }
        stale = surface?.source.isStale ?? false
        rawFeatureCount = surface?.source.rawElementCount ?? 0
        runwayCount = surface?.runways.count ?? 0
        taxiwayCount = surface?.taxiwaysOnly.count ?? 0
        taxilaneCount = surface?.taxilanes.count ?? 0
        holdingPositionCount = surface?.holdingPositions.count ?? 0
        parkingCount = surface?.parkingPositions.count ?? 0
        apronCount = surface?.aprons.count ?? 0
        graphNodeCount = graph?.nodes.count ?? 0
        graphEdgeCount = graph?.edges.count ?? 0
        disconnectedComponents = max(0, (graph?.componentCount ?? 1) - 1)
        inferredConnectors = graph?.inferredConnectorCount ?? 0
        self.snappedSegment = snappedSegment
        self.datasetConfidence = datasetConfidence
        self.routeConfidence = routeConfidence

        if let route {
            routeSummary = route.isDeparture
                ? "Departure → \(route.destinationLabel) via \(route.taxiwaysText)"
                : "Arrival → \(route.destinationLabel) via \(route.taxiwaysText)"
            routeDistanceMeters = route.distanceMeters
            runwayCrossings = route.crossings.count
        } else {
            routeSummary = "No route calculated"
            routeDistanceMeters = nil
            runwayCrossings = 0
        }

        if let c = activeCrossing {
            nextCrossing = "Runway \(c.runwayIdent) (\(c.confidence.title))"
        } else {
            nextCrossing = "—"
        }
        self.crossingState = crossingState.title
        if awaitingCrossingReadback {
            authorizationState = "Awaiting read back"
        } else if let idx = authorizedCrossingIndex {
            authorizationState = "Authorized (crossing \(idx + 1))"
        } else {
            authorizationState = crossingState.isAuthorized ? "Authorized" : "Not authorized"
        }

        switch status {
        case .idle: statusText = "Idle"
        case .loading: statusText = "Loading"
        case .ready: statusText = "Ready"
        case .unavailable(let reason): statusText = "Unavailable — \(reason)"
        case .error(let reason): statusText = "Error — \(reason)"
        }
        self.lastError = lastError
    }

    /// Shareable plain-text export of the surface diagnostics.
    func exportText() -> String {
        func date(_ d: Date?) -> String {
            guard let d else { return "—" }
            let f = ISO8601DateFormatter()
            return f.string(from: d)
        }
        var lines: [String] = []
        lines.append("IFATC Companion — Airport Surface Diagnostics")
        lines.append("Airport: \(airportID)")
        lines.append("Source: \(sourceProvider)")
        lines.append("License: \(license)")
        lines.append("Attribution: \(attribution)")
        lines.append("Query endpoint: \(endpoint)")
        lines.append("Fetch date: \(date(fetchDate))")
        lines.append("Cache age: \(cacheAgeDays.map { "\($0) days" } ?? "—")\(stale ? " (stale)" : "")")
        lines.append("Source features: \(rawFeatureCount)")
        lines.append("Runways: \(runwayCount)")
        lines.append("Taxiways: \(taxiwayCount)")
        lines.append("Taxilanes: \(taxilaneCount)")
        lines.append("Holding positions: \(holdingPositionCount)")
        lines.append("Gates/parking: \(parkingCount)")
        lines.append("Aprons: \(apronCount)")
        lines.append("Graph nodes: \(graphNodeCount)")
        lines.append("Graph edges: \(graphEdgeCount)")
        lines.append("Disconnected components: \(disconnectedComponents)")
        lines.append("Inferred connectors: \(inferredConnectors)")
        lines.append("Snapped segment: \(snappedSegment)")
        lines.append("Route: \(routeSummary)")
        lines.append("Route distance: \(routeDistanceMeters.map { "\(Int($0)) m" } ?? "—")")
        lines.append("Runway crossings: \(runwayCrossings)")
        lines.append("Dataset confidence: \(datasetConfidence.title)")
        lines.append("Route confidence: \(routeConfidence.title)")
        lines.append("Next crossing: \(nextCrossing)")
        lines.append("Crossing state: \(crossingState)")
        lines.append("Authorization: \(authorizationState)")
        lines.append("Status: \(statusText)")
        lines.append("Last error: \(lastError ?? "—")")
        return lines.joined(separator: "\n")
    }
}

File: Android/core/src/main/kotlin/com/h3consultingpartners/ifatccompanion/core/surface/AirportSurfaceDiagnostics.kt
package com.h3consultingpartners.ifatccompanion.core.surface

import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import kotlin.math.max

/**
 * The status line the surface coordinator is in, spelled once here so the wording
 * cannot drift between the coordinator and this snapshot.
 *
 * CONTRACT for the taxi/coordinator agent: `AirportSurfaceStatus` (a sealed type with
 * `idle`/`loading`/`ready`/`unavailable(reason)`/`error(reason)`) lives in that agent's
 * package; it maps its own case to one of these strings and passes the result to
 * [AirportSurfaceDiagnostics.from] as `statusText`.
 */
object AirportSurfaceStatusText {
    const val IDLE = "Idle"
    const val LOADING = "Loading"
    const val READY = "Ready"
    fun unavailable(reason: String): String = "Unavailable — $reason"
    fun error(reason: String): String = "Error — $reason"
}

/**
 * The counts this snapshot needs from the connected surface graph.
 *
 * CONTRACT: `SurfaceGraph` belongs to the graph/routing agent. Rather than depend on it
 * (or duplicate it), the coordinator hands over this digest — null when no graph has
 * been built, which reads as zero nodes/edges and one component, exactly as the Swift's
 * `graph?.componentCount ?? 1` does.
 */
data class SurfaceGraphSummary(
    val nodeCount: Int,
    val edgeCount: Int,
    val componentCount: Int,
    val inferredConnectorCount: Int,
)

/**
 * The bits of a calculated taxi route this snapshot needs.
 *
 * CONTRACT: `SurfaceTaxiRoute` belongs to the taxi-routing agent; the coordinator maps
 * one into this digest. Null means no route has been calculated.
 */
data class TaxiRouteSummary(
    val isDeparture: Boolean,
    val destinationLabel: String,
    val taxiwaysText: String,
    val distanceMeters: Double,
    val crossingCount: Int,
)

/**
 * The active runway crossing, as this snapshot reports it.
 *
 * CONTRACT: `RouteCrossing` belongs to the taxi-routing agent; the coordinator maps the
 * active crossing (if any) into this digest.
 */
data class ActiveCrossingSummary(
    val runwayIdent: String,
    val confidence: SurfaceConfidence,
)

/**
 * A point-in-time snapshot of the airport-surface feature for the Airport Surface
 * Diagnostics view and its text export. Always identifies OpenStreetMap / ODbL 1.0 and
 * carries the visible attribution.
 *
 * Ported from `IFATCCompanion/AirportSurface/AirportSurfaceDiagnostics.swift`. The Swift
 * initializer takes the coordinator's own graph/route/crossing/status types; the ones
 * this package does not own arrive as the digests above (see each CONTRACT note). The
 * Swift `kind: TaxiKind` and `progress: RouteTracker.Progress?` parameters are not taken
 * here because the initializer never reads either of them.
 */
data class AirportSurfaceDiagnostics(
    val airportID: String,
    val sourceProvider: String,
    val license: String,
    val attribution: String,
    val endpoint: String,
    /** Epoch millis the extract was fetched, or null when there is no surface loaded. */
    val fetchDateMillis: Long?,
    val cacheAgeDays: Int?,
    val stale: Boolean,
    val rawFeatureCount: Int,
    val runwayCount: Int,
    val taxiwayCount: Int,
    val taxilaneCount: Int,
    val holdingPositionCount: Int,
    val parkingCount: Int,
    val apronCount: Int,
    val buildingCount: Int,
    val schemaVersion: Int,
    val schemaOutdated: Boolean,
    val graphNodeCount: Int,
    val graphEdgeCount: Int,
    val disconnectedComponents: Int,
    val inferredConnectors: Int,
    val snappedSegment: String,
    val routeSummary: String,
    val routeDistanceMeters: Double?,
    val runwayCrossings: Int,
    val routeConfidence: SurfaceConfidence,
    val datasetConfidence: SurfaceConfidence,
    val nextCrossing: String,
    val crossingState: String,
    val authorizationState: String,
    val statusText: String,
    val lastError: String?,
) {

    /** Shareable plain-text export of the surface diagnostics. */
    fun exportText(): String {
        val lines = mutableListOf<String>()
        lines.add("IFATC Companion — Airport Surface Diagnostics")
        lines.add("Airport: $airportID")
        lines.add("Source: $sourceProvider")
        lines.add("License: $license")
        lines.add("Attribution: $attribution")
        lines.add("Query endpoint: $endpoint")
        lines.add("Fetch date: ${formatDate(fetchDateMillis)}")
        lines.add(
            "Cache age: ${cacheAgeDays?.let { "$it days" } ?: "—"}" +
                if (stale) " (stale)" else "",
        )
        lines.add("Source features: $rawFeatureCount")
        lines.add("Runways: $runwayCount")
        lines.add("Taxiways: $taxiwayCount")
        lines.add("Taxilanes: $taxilaneCount")
        lines.add("Holding positions: $holdingPositionCount")
        lines.add("Gates/parking: $parkingCount")
        lines.add("Aprons: $apronCount")
        lines.add("Buildings/terminals: $buildingCount")
        lines.add(
            "Schema version: $schemaVersion" +
                if (schemaOutdated) " (outdated — will refresh)" else "",
        )
        lines.add("Graph nodes: $graphNodeCount")
        lines.add("Graph edges: $graphEdgeCount")
        lines.add("Disconnected components: $disconnectedComponents")
        lines.add("Inferred connectors: $inferredConnectors")
        lines.add("Snapped segment: $snappedSegment")
        lines.add("Route: $routeSummary")
        lines.add("Route distance: ${routeDistanceMeters?.let { "${it.toInt()} m" } ?: "—"}")
        lines.add("Runway crossings: $runwayCrossings")
        lines.add("Dataset confidence: ${datasetConfidence.title}")
        lines.add("Route confidence: ${routeConfidence.title}")
        lines.add("Next crossing: $nextCrossing")
        lines.add("Crossing state: $crossingState")
        lines.add("Authorization: $authorizationState")
        lines.add("Status: $statusText")
        lines.add("Last error: ${lastError ?: "—"}")
        return lines.joinToString("\n")
    }

    companion object {
        /** `ISO8601DateFormatter().string(from:)` — an instant in UTC, second precision. */
        private val iso8601: DateTimeFormatter =
            DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'").withZone(ZoneOffset.UTC)

        private fun formatDate(millis: Long?): String =
            if (millis == null) "—" else iso8601.format(Instant.ofEpochMilli(millis))

        /**
         * Build a snapshot. Mirrors the Swift `init(surface:graph:route:…)`, including the
         * fall-backs it applies when there is no surface loaded yet.
         *
         * [crossingStateTitle] and [crossingStateAuthorized] are `RunwayCrossingState.title`
         * and `.isAuthorized` — that enum belongs to the taxi/coordinator agent.
         */
        fun from(
            surface: AirportSurfaceModel?,
            graph: SurfaceGraphSummary?,
            route: TaxiRouteSummary?,
            statusText: String,
            datasetConfidence: SurfaceConfidence,
            routeConfidence: SurfaceConfidence,
            crossingStateTitle: String,
            crossingStateAuthorized: Boolean,
            activeCrossing: ActiveCrossingSummary?,
            awaitingCrossingReadback: Boolean,
            authorizedCrossingIndex: Int?,
            snappedSegment: String,
            lastError: String?,
            nowMillis: Long = System.currentTimeMillis(),
        ): AirportSurfaceDiagnostics {
            val routeSummaryText: String
            val routeDistance: Double?
            val crossings: Int
            if (route != null) {
                routeSummaryText = if (route.isDeparture) {
                    "Departure → ${route.destinationLabel} via ${route.taxiwaysText}"
                } else {
                    "Arrival → ${route.destinationLabel} via ${route.taxiwaysText}"
                }
                routeDistance = route.distanceMeters
                crossings = route.crossingCount
            } else {
                routeSummaryText = "No route calculated"
                routeDistance = null
                crossings = 0
            }

            val nextCrossing = if (activeCrossing != null) {
                "Runway ${activeCrossing.runwayIdent} (${activeCrossing.confidence.title})"
            } else {
                "—"
            }

            val authorizationState = when {
                awaitingCrossingReadback -> "Awaiting read back"
                authorizedCrossingIndex != null -> "Authorized (crossing ${authorizedCrossingIndex + 1})"
                crossingStateAuthorized -> "Authorized"
                else -> "Not authorized"
            }

            return AirportSurfaceDiagnostics(
                airportID = surface?.icao ?: "—",
                sourceProvider = OSMSurface.PROVIDER_NAME,
                license = OSMSurface.LICENSE_SHORT_NAME,
                attribution = OSMSurface.ATTRIBUTION_TEXT,
                endpoint = surface?.source?.endpoint ?: OSMSurface.PRIMARY_OVERPASS_ENDPOINT,
                fetchDateMillis = surface?.source?.fetchDateMillis,
                cacheAgeDays = surface?.source?.cacheAgeDays(nowMillis),
                stale = surface?.source?.isStale(nowMillis) ?: false,
                rawFeatureCount = surface?.source?.rawElementCount ?: 0,
                runwayCount = surface?.runways?.size ?: 0,
                taxiwayCount = surface?.taxiwaysOnly?.size ?: 0,
                taxilaneCount = surface?.taxilanes?.size ?: 0,
                holdingPositionCount = surface?.holdingPositions?.size ?: 0,
                parkingCount = surface?.parkingPositions?.size ?: 0,
                apronCount = surface?.aprons?.size ?: 0,
                buildingCount = surface?.buildings?.size ?: 0,
                schemaVersion = surface?.source?.schemaVersion ?: OSMSurface.SURFACE_SCHEMA_VERSION,
                schemaOutdated = surface?.source?.isOutdatedSchema ?: false,
                graphNodeCount = graph?.nodeCount ?: 0,
                graphEdgeCount = graph?.edgeCount ?: 0,
                disconnectedComponents = max(0, (graph?.componentCount ?: 1) - 1),
                inferredConnectors = graph?.inferredConnectorCount ?: 0,
                snappedSegment = snappedSegment,
                routeSummary = routeSummaryText,
                routeDistanceMeters = routeDistance,
                runwayCrossings = crossings,
                routeConfidence = routeConfidence,
                datasetConfidence = datasetConfidence,
                nextCrossing = nextCrossing,
                crossingState = crossingStateTitle,
                authorizationState = authorizationState,
                statusText = statusText,
                lastError = lastError,
            )
        }
    }
}
